File: External References/Devart/dotConnect/Linq/Samples/LinqConnect/VB/ConnectionDialog.vb
Imports System.Text
Imports System.Configuration
Imports System.Data.Common

Public Class ConnectionDialog

  Public parameters As Dictionary(Of String, String) = New Dictionary(Of String, String)
  Private Shared providerNames As String()
  Public providerPrefix As String
  Private Shared providerPrefixes As String()

  Shared Sub New()
        ConnectionDialog.providerNames = New String() {"Devart.Data.Oracle", "Devart.Data.MySql", "Devart.Data.PostgreSql", "Devart.Data.SQLite", "System.Data.SqlClient"}
        ConnectionDialog.providerPrefixes = New String() {"Oracle", "MySql", "PostgreSql", "SQLite", "SqlClient"}
  End Sub

  Private Sub ConnectDialog_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

    Dim table As DataTable = DbProviderFactories.GetFactoryClasses
    If (Not table Is Nothing) Then
      Dim providerName As String
      For Each providerName In ConnectionDialog.providerNames
        Dim row As DataRow = table.Rows.Find(providerName)
        If (Not row Is Nothing) Then
          Me.cbProviders.Items.Add(row.Item(0))
        End If
      Next
      If (Me.cbProviders.Items.Count > 0) Then
        Me.cbProviders.SelectedItem = Me.cbProviders.Items.Item(0)
      Else
        Me.bbNoSupportedFound.Visible = True
      End If
      If (Me.cbProviders.Items.Count = 1) Then
        Me.cbProviders.Enabled = False
      End If
    End If
  End Sub

  Public Shared Function Show(ByRef providerConnectionString As String, ByRef providerPrefix As String) As Boolean
    Dim form As New ConnectionDialog
    If (form.ShowDialog <> DialogResult.OK) Then
      providerConnectionString = Nothing
      providerPrefix = Nothing
      Return False
    End If
    providerPrefix = form.providerPrefix
    Dim sb As New StringBuilder
    Dim index As Integer = 0
    Dim comp As Control
    For Each comp In form.panel1.Controls
      If (comp.Tag Is Nothing) Then
        Continue For
      End If
      Dim key As String = CStr(comp.Tag)
      Dim text As String = String.Empty
      If TypeOf comp Is TextBox Then
        [text] = TryCast(comp, TextBox).Text
      ElseIf TypeOf comp Is CheckBox Then
        [text] = IIf(TryCast(comp, CheckBox).Checked, "True", "False")
      ElseIf TypeOf comp Is NumericUpDown Then
        [text] = TryCast(comp, NumericUpDown).Value.ToString
      End If
      sb.AppendFormat("{0}={1};", key, [text])
      index += 1
    Next
    providerConnectionString = sb.ToString
    Return True
  End Function

  Private Sub cbProviders_SelectedIndexChanged(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles cbProviders.SelectedIndexChanged

    Dim providerIndex As Integer = -1
    Dim i As Integer
    For i = 0 To ConnectionDialog.providerNames.Length - 1
      If (CStr(Me.cbProviders.SelectedItem).IndexOf(ConnectionDialog.providerPrefixes(i), StringComparison.InvariantCultureIgnoreCase) >= 0) Then
        providerIndex = i
        Exit For
      End If
    Next i
    If (providerIndex >= 0) Then
      Me.providerPrefix = ConnectionDialog.providerPrefixes(providerIndex)
      Dim connectionString As String = ConfigurationManager.ConnectionStrings.Item(("ConnectionString" & Me.providerPrefix)).ToString
      Me.InitConnectionStringControls(connectionString)
    End If
  End Sub

  Private Sub InitConnectionStringControls(ByVal providerConnectionString As String)

    Me.panel1.Controls.Clear()
    Me.parameters.Clear()
    Dim parts As String() = providerConnectionString.Split(New Char() {"="c, ";"c})
    If ((parts.Length Mod 2) = 0) Then
      Dim keyCount As Integer = (parts.Length / 2)
      Dim i As Integer
      For i = 0 To keyCount - 1
        Me.parameters.Add(parts((i * 2)), parts(((i * 2) + 1)))
      Next i
      Dim shiftY As Integer = 10
      Dim index As Integer = 1
      Dim components As New List(Of Control)
      Dim key As String
      For Each key In Me.parameters.Keys
        Dim comp As Control
        Dim intValue As Long
        Dim label As New Label
        label.Location = New Point(12, shiftY)
        label.Size = New Size(&H4B, &H13)
        label.Text = key
        label.TextAlign = ContentAlignment.MiddleLeft
        Dim value As String = Me.parameters.Item(key)
        If Long.TryParse(value, intValue) Then
          Dim numeric As New NumericUpDown
          numeric.Maximum = 65535
          numeric.Value = intValue
          comp = numeric
        Else
          Dim boolValue As Boolean
          If Boolean.TryParse(value, boolValue) Then
            Dim checkbox As New CheckBox
            checkbox.Checked = boolValue
            comp = checkbox
          Else
            Dim textBox As New TextBox
            If key.ToLower.Contains("password") Then
              textBox.PasswordChar = "*"c
            End If
            textBox.Text = value
            comp = textBox
          End If
        End If
        comp.Location = New Point(90, shiftY)
        comp.Size = New Size(&HBB, 20)
        comp.TabIndex = (index + 1)
        comp.Tag = key
        Me.panel1.Controls.Add(comp)
        Me.panel1.Controls.Add(label)
        components.Add(comp)
        shiftY = (shiftY + &H19)
        index += 1
      Next
    End If
  End Sub

End Class

File: External References/Devart/dotConnect/Linq/Samples/LinqConnect/CS/ConnectionDialog.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Data.Common;
using System.Configuration;

namespace LinqDemo {

  public partial class ConnectionDialog : Form {

    private static string[] providerNames = new string[] { "Devart.Data.Oracle", "Devart.Data.MySql", "Devart.Data.PostgreSql", "Devart.Data.SQLite", "System.Data.SqlClient" };
    private static string[] providerPrefixes = new string[] { "Oracle", "MySql", "PostgreSql", "SQLite", "SqlClient" };

    public Dictionary<string, string> parameters = new Dictionary<string, string>();
    public string providerPrefix;

    public ConnectionDialog() {
      InitializeComponent();
    }

    private void ConnectionDialog_Load(object sender, EventArgs e) {

      DataTable table = DbProviderFactories.GetFactoryClasses();
      if (table != null) {
        foreach (string providerName in providerNames) {
          DataRow row = table.Rows.Find(providerName);
          if (row != null) {
            cbProviders.Items.Add(row[0]);
          }
        }
        if (cbProviders.Items.Count > 0)
          cbProviders.SelectedItem = cbProviders.Items[0];
        else
          bbNoSupportedFound.Visible = true;
        if (cbProviders.Items.Count == 1)
          cbProviders.Enabled = false;
      }
    }

    private void cbProviders_SelectedIndexChanged(object sender, EventArgs e) {
      
      int providerIndex = -1;
      for (int i = 0; i < providerNames.Length;i++) {
        if (((string)cbProviders.SelectedItem).IndexOf(providerPrefixes[i], StringComparison.InvariantCultureIgnoreCase) >= 0) {
          providerIndex = i;
          break;
        }
      }
      if (providerIndex >= 0) {
        providerPrefix = providerPrefixes[providerIndex];
        string connectionString = ConfigurationManager.ConnectionStrings["ConnectionString" + providerPrefix].ToString();
        InitConnectionStringControls(connectionString);
      }
    }

    private void InitConnectionStringControls(string providerConnectionString) {

      panel1.Controls.Clear();
      parameters.Clear();

      string[] parts = providerConnectionString.Split('=', ';');
      if (parts.Length % 2 != 0)
        return;
      int keyCount = parts.Length / 2;
      for (int i = 0; i < keyCount; i++)
        parameters.Add(parts[i * 2], parts[i * 2 + 1]);

      // show Connection Dialog
      int shiftY = 10;
      int index = 1;
      List<Control> components = new List<Control>();
      foreach (string key in parameters.Keys) {

        // Key
        Label label = new Label();
        label.Location = new Point(12, shiftY);
        label.Size = new Size(75, 19);
        label.Text = key;
        label.TextAlign = ContentAlignment.MiddleLeft;

        // Value
        Control comp;
        string value = parameters[key];
        Int64 intValue;
        bool boolValue;
        if (Int64.TryParse(value, out intValue)) {
          NumericUpDown numeric = new NumericUpDown();
          numeric.Maximum = 65535;
          numeric.Value = intValue;
          comp = numeric;
        }
        else
          if (Boolean.TryParse(value, out boolValue)) {
            CheckBox checkbox = new CheckBox();
            checkbox.Checked = boolValue;
            comp = checkbox;
          }
          else {
            TextBox textBox = new TextBox();
            if (key.ToLower().Contains("password"))
              textBox.PasswordChar = '*';
            textBox.Text = value;
            comp = textBox;
          }
        comp.Location = new Point(90, shiftY);
        comp.Size = new Size(187, 20);
        comp.TabIndex = index + 1;
        comp.Tag = key;

        panel1.Controls.Add(comp);
        panel1.Controls.Add(label);
        components.Add(comp);
        shiftY += 25;
        index++;
      }
    }

    public static bool Show(out string providerConnectionString, out string providerPrefix) {
      
      ConnectionDialog form = new ConnectionDialog();
      if (form.ShowDialog() != DialogResult.OK) {
        providerConnectionString = null;
        providerPrefix = null;
        return false;
      }

      providerPrefix = form.providerPrefix;

      // new connection string
      StringBuilder sb = new StringBuilder();
      int index = 0;
      foreach (Control comp in form.panel1.Controls) {
        if (comp.Tag == null)
          continue;
        string key = (string)comp.Tag;
        string text = String.Empty;
        if (comp is TextBox)
          text = (comp as TextBox).Text;
        else
          if (comp is CheckBox)
            text = (comp as CheckBox).Checked ? "True" : "False";
          else
            if (comp is NumericUpDown)
              text = (comp as NumericUpDown).Value.ToString();
        sb.AppendFormat("{0}={1};", key, text);
        index++;
      }
      providerConnectionString = sb.ToString();
      return true;
    }

  }
}
